### ISSUE TITLE
Support for FontAwesome

### ISSUE THREAD

**Issue by xclayl** [stephenh/tessell#21]
When we put fontawesome, http://fortawesome.github.io/Font-Awesome/, in our resources folder, the @Source() attributes are incorrect: @Source("css\font-awesome.gen.css") for example -- ie the backslash is not escaped.  
Even if this were correct, you get the following error when running the app (super dev mode)
[ERROR] Could not find no-arg method named fontawesomeWebfontEotV43Data in type project.client.resources.AppResources

**Comment by stephenh** [comment]
Hm. The backslash is odd, I'll take a look at that...

My current guess is that the font-awesome CSS file assumes it's various font files will live next to it (e.g. in a traditional webapp.com/css/fontawesome.css, webapp.com/fonts/fontawesome.svg) and that putting it into a ClientBundle is confusing it.

**Comment by stephenh** [comment]
Okay, I pushed a branch that makes font awesome load/render correctly in the tessell-examples app.

I have not merged this to master as, per the commit message, I had to differentiate between "font.svg" and "font.eot" by making methods like fontSvgData and fontEotData(). This is a breaking change for apps that don't have file name collisions, and ideally this would be conditional.

I can poke some more at that, but you can see if this works for you. To try it out you can run "./gradlew install" to install a snapshot into your ~/.m2 repository.

Also note that I branched of Tessell 3.x, which has switched to Java 8/GWT snapshot builds. If you're not ready for that, you can create a branch off the pre-3.x changes and cherry pick the 21-support-awesome-branch changes on to it. There should not be any conflicts.

**Comment by xclayl** [comment]
I'm not familar with gradle and git so I'll try to get up to speed this weekend and try it out.  I see the branch in github so I should be able to download it.  Thanks for looking into it so quickly!

**Comment by stephenh** [comment]
Another option (that is actually what we use in our app for bootstrap+font-awesome because GWT's CSS parser, until GSS, couldn't even parse some CSS3 things in bootstrap) is that instead of putting the real CSS file into the GWT app, and having it end up in the ClientBundle/etc., you can generate a "placeholder" css file:

https://gist.github.com/stephenh/3d871a6ad9bd743ff01e

Put this into your resources directory, and you'll get a FontAwesomeStyle with all the .faFoo(), .faBar(), etc.

Of course, there is not actual content in the placeholder file, it just lets GWT know about the CSS class names. So, because there is no content, you don't have to worry about font files/resources/etc. lining up.

Then in your app.html file, you use the regular font-awesome installation, e.g. use a link tag and pulls it into the app.html file just like a normal/non-GWT app.

This worked well for us, we just have to generate a new "placeholder" file when we add new CSS files to the real bootstrap CSS file (that we've forked a bit).

It would still be nice for Tessell to just handle this though, and I've got it mostly working, but just wanted to mention the option.

**Comment by stephenh** [comment]
I've pushed a fix for Source("css\foo.css") vs. Source("css/foo.css") to master and released it as tessell-user/-dev 3.0.5. Let me know if that bug is fixed.

**Comment by xclayl** [comment]
It looks like Tessell v3.x is on Java 8 which I assume is a beta release of GWT 2.8.  The GWT version we use supports Java 7.  I think we're out of luck for the moment.

A colleague also suggested the CSS stubs technique and I think we'll go down this path for now. Thanks for the work around suggestion.

I noticed that the code changes convert "\" to "/" which would work for us (I tried it by manually changing the @Source tag).  However if you have two files with the same name, but different directories, the generated code gets mixed up. For example it generates the following:

    @Source("bg.jpg")
    @Source("sub\bg.jpg")
    public ImageResource bg();

    @Source("bg.jpg")
    @Source("sub\bg.jpg")
    public DataResource bgData();

I'm just mementioning this in case you wanted to tackle the name conflict issue (file extension conflict)

Thanks for all you've done. We really love Tessell overall.

**Comment by stephenh** [comment]
> Tessell v3 == Java 8

Yes, that's true... For just the back/forward slash issue, I cherry picked the change to a tessell-2 branch; version 2.20.1 should be in the Tessell repo soon.

> two files with same name but different directories

...yeah, I've known about that since I originally wrote the generator :-), but I wasn't really sure how to handle it nicely. I've filed #23 to follow up on that.

I've also filed #22 to more directly address resources with different extensions.

**Comment by stephenh** [comment]
> Thanks for all you've done. We really love Tessell overall.

Forgot to say thanks for that; I really appreciate it.
